proxy/coco/src/state.rs

```diff
@@ -0,0 +1,733 @@
+    pub fn init_owner(&self, signer: &signer::BoxedSigner, handle: &str) -> Result<User, Error> {
+        let user = self.init_user(signer, handle)?;
+        let user = verify_user(user)?;
+
+        self.set_default_owner(user.clone())?;
+
+        Ok(user)
+    }
+
+    /// Given some hints as to where you might find it, get the urn of the project found at `url`.
+    ///
+    /// **N.B.** This needs to be run with `tokio::spawn_blocking`.
+    ///
+    /// # Errors
+    ///   * Could not successfully acquire a lock to the API.
+    ///   * Could not open librad storage.
+    ///   * Failed to clone the project.
+    ///   * Failed to set the rad/self of this project.
+    pub fn clone_project<Addrs>(&self, url: RadUrl, addr_hints: Addrs) -> Result<RadUrn, Error>
+    where
+        Addrs: IntoIterator<Item = SocketAddr>,
+    {
+        let storage = self.api.storage().reopen()?;
+        let repo = storage.clone_repo::<librad_project::ProjectInfo, _>(url, addr_hints)?;
+        repo.set_rad_self(storage::RadSelfSpec::Default)?;
+        Ok(repo.urn)
+    }
+
+    /// Get the project found at `urn`.
+    ///
+    /// # Errors
+    ///
+    ///   * Resolving the project fails.
+    pub fn get_project<P>(
+        &self,
+        urn: &RadUrn,
+        peer: P,
+    ) -> Result<librad_project::Project<entity::Draft>, Error>
+    where
+        P: Into<Option<PeerId>>,
+    {
+        let storage = self.api.storage().reopen()?;
+
+        Ok(storage.metadata_of(urn, peer)?)
+    }
+
+    /// Returns the list of [`librad_project::Project`]s for your peer.
+    ///
+    /// # Errors
+    ///
+    ///   * Retrieving the project entities from the store fails.
+    #[allow(
+        clippy::match_wildcard_for_single_variants,
+        clippy::wildcard_enum_match_arm
+    )]
+    pub fn list_projects(&self) -> Result<Vec<librad_project::Project<entity::Draft>>, Error> {
+        let project_meta = {
+            let storage = self.api.storage().reopen()?;
+            let owner = storage.default_rad_self()?;
+
+            let meta = storage.all_metadata()?;
+            meta.flat_map(|entity| {
+                let entity = entity.ok()?;
+                let rad_self = storage.get_rad_self(&entity.urn()).ok()?;
+
+                // We only list projects that are owned by the peer
+                if rad_self.urn() != owner.urn() {
+                    return None;
+                }
+
+                entity.try_map(|info| match info {
+                    entity::data::EntityInfo::Project(info) => Some(info),
+                    _ => None,
+                })
+            })
+            .collect::<Vec<_>>()
+        };
+
+        Ok(project_meta)
+    }
+
+    /// Retrieves the [`librad::git::refs::Refs`] for the state owner.
+    ///
+    /// # Errors
+    ///
+    /// * if opening the storage fails
+    pub fn list_owner_project_refs(&self, urn: &RadUrn) -> Result<Refs, Error> {
```

Comment:
Can we get a patch for this @xla /cc @sarahscott 

It's causing Sarah issues on her work :)

Reply:
#891